Source/WebKit/UIProcess/API/C/curl/WKWebsiteDataStoreConfigurationRefCurl.h

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2023 Sony Interactive Entertainment Inc.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <WebKit/WKBase.h>
+#include <WebKit/WKWebsiteDataStoreConfigurationRef.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+WK_EXPORT WKStringRef WKWebsiteDataStoreConfigurationCopyAlternativeServicesStorageFile(WKWebsiteDataStoreConfigurationRef configuration);
+WK_EXPORT void WKWebsiteDataStoreConfigurationSetAlternativeServicesStorageFile(WKWebsiteDataStoreConfigurationRef configuration, WKStringRef cacheFile);
```

Comment:
Do applications have to use this API to enable Alt-Svc?
Do you really need to set a filename?
What about an idea of using an existing API `WKContextConfigurationSetDiskCacheDirectory` and a fixed file name "altsvc-cache.txt"?

Reply:
> Do applications have to use this API to enable Alt-Svc?

Some applications may not require Alt-Svc. Therefore I would like to be able to enable/disable this feature.

> Do you really need to set a filename?
> What about an idea of using an existing API WKContextConfigurationSetDiskCacheDirectory and a fixed file name "altsvc-cache.txt"?

This is misleading to the user as there is no relationship between Disk Cache and Alt-Svc.

Reply:
Would you like to set a default filename?

For example:
https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h#L371
https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp#L2038-L2046
https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp#L92

Reply:
I've tried creating a patch that sets the default value, and I think this is the simplest way to do it.

[default.txt](https://github.com/WebKit/WebKit/files/11156148/default.txt)

Reply:
Fixed to set default filename.

Reply:
I think Alt-Svc cache is a disk cache of network response.
How other browsers do? Do they have a UI to clear Alt-Svc cache? How to clear Alt-Svc cache?
How to enable/disable Alt-Svc with the new API? Call WKWebsiteDataStoreConfigurationSetAlternativeServicesStorageFile with empty string to disable? It's disabled by default only for PlayStation port?
How to clear Alt-Svc cache? Remove the file by using a system API?
Why do you need enable/disable Alt-Svc? Do you need API to enable/disable other HTTP features?

Reply:
> How other browsers do?

The Cocoa port is designed to specify the Alt-Svc storage directory.
https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h#L97

> Do they have a UI to clear Alt-Svc cache?

Delete it as part of the WebsiteData.

> How to clear Alt-Svc cache?
> Remove the file by using a system API?

For Cocoa port, it will be removed using the removeDataOfTypes function.
https://github.com/WebKit/WebKit/blob/main/Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm#L706

In the case of Curl port, it is necessary to prepare a dedicated API like PR below.
https://github.com/WebKit/WebKit/pull/6200

However, as Alex mentioned in this PR, it would be better to have an API like removeDataOfTypes rather than writing a dedicated function.

Reply:
> Call WKWebsiteDataStoreConfigurationSetAlternativeServicesStorageFile with empty string to disable?

Hmm.. If WKWebsiteDataStoreConfigurationSetAlternativeServicesStorageFile is an empty string, it will not be possible to overwrite its value. This means that once the default value is set it cannot be turned off.

 https://github.com/WebKit/WebKit/pull/12404/files#diff-f5dc1cba4ced1243c9430dea1be6d75874b4b8d452fd989cf498d3810c164198R382

Reply:
> However, as Alex mentioned in this PR, it would be better to have an API like removeDataOfTypes
> rather than writing a dedicated API. As such, I did not provide an API for removal in this PR.

Since such APIs have not been created yet, I have prepared a dedicated API `WKWebsiteDataStoreRemoveAlternativeServicesStorageFile`.

Reply:
Removed WKWebsiteDataStoreConfigurationSetAlternativeServicesStorageFile.